exp/runtime/hooks/api/v1alpha1/lifecyclehooks_types.go

```diff
@@ -23,59 +23,60 @@ import (
 	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
 )
 
-// BlockingResponse is the response of a blocking lifecycle hook.
+// BeforeClusterCreate Hook
+
+// BeforeClusterCreateRequest is the request of the hook.
 // +kubebuilder:object:root=true
-type BlockingResponse struct {
+type BeforeClusterCreateRequest struct {
 	metav1.TypeMeta `json:",inline"`
 
-	// Status of the call. One of "Success" or "Failure".
-	Status ResponseStatus `json:"status"`
-
-	// RetryAfterSeconds when set to a non-zero signifies that the hook
-	// needs to be retried at a future time.
-	RetryAfterSeconds int `json:"retryAfterSeconds"`
-
-	// A human-readable description of the status of the call.
-	Message string `json:"message"`
+	// The cluster object the lifecycle hook corresponds to.
+	Cluster clusterv1.Cluster `json:"cluster"`
 }
 
-// NonBlockingResponse is the response of a non-blocking lifecycle hook.
+// BeforeClusterCreateResponse is the response of BeforeClusterCreate hook.
 // +kubebuilder:object:root=true
-type NonBlockingResponse struct {
+type BeforeClusterCreateResponse struct {
 	metav1.TypeMeta `json:",inline"`
 
 	// Status of the call. One of "Success" or "Failure".
 	Status ResponseStatus `json:"status"`
 
+	// RetryAfterSeconds when set to a non-zero signifies that the hook
+	// will be called again at a future time.
+	RetryAfterSeconds int `json:"retryAfterSeconds"`
+
 	// A human-readable description of the status of the call.
 	Message string `json:"message"`
 }
 
-// BeforeClusterCreate Hook
+func BeforeClusterCreate(*BeforeClusterCreateRequest, *BeforeClusterCreateResponse) {}
 
-// BeforeClusterCreateRequest is the request of the hook.
+// AfterControlPlaneInitialized Hook
+
+// AfterControlPlaneInitializedRequest is the request of the hook.
 // +kubebuilder:object:root=true
-type BeforeClusterCreateRequest struct {
+type AfterControlPlaneInitializedRequest struct {
 	metav1.TypeMeta `json:",inline"`
 
 	// The cluster object the lifecycle hook corresponds to.
 	Cluster clusterv1.Cluster `json:"cluster"`
 }
 
-func BeforeClusterCreate(*BeforeClusterCreateRequest, *BlockingResponse) {}
-
-// AfterControlPlaneInitialized Hook
-
-// AfterControlPlaneInitializedRequest is the request of the hook.
+// AfterControlPlaneInitializedResponse is the response of AfterControlPlaneInitialized hook.
 // +kubebuilder:object:root=true
-type AfterControlPlaneInitializedRequest struct {
+type AfterControlPlaneInitializedResponse struct {
 	metav1.TypeMeta `json:",inline"`
 
-	// The cluster object the lifecycle hook corresponds to.
-	Cluster clusterv1.Cluster `json:"cluster"`
+	// Status of the call. One of "Success" or "Failure".
+	Status ResponseStatus `json:"status"`
```

Comment:
Maybe a dumb question and unrelated to this PR (because it already was like this with the old code):

The `AfterControlPlaneInitializedResponse` has no `RetryAfterSeconds` which (if I got it right) means that this can't be a blocking response. (same for `AfterClusterUpgradeResponse`).

Why is that? I can think of a cluster-operator wanting his gitops stuff provision stuff and block until that finished.

Reply:
Makes sense 👍 
Additional question: this hook is then not blocking, will it still get retried later on until it succeeds or will it stay in failed state?

Reply:
If there's an error from this hook it will return an error in the reconcile, so in reality it will retry until it returns a non-error. The idea here is to not have a best-effort hook that might not be called. 

An error will stop all reconciliation though, and the guideline is to only use it if this is an actual error state, not to simply block where you otherwise could not.

Reply:
Sound good, thanks for clarifiying!